core/lib/multivm/src/versions/vm_fast/evm_deploy_tracer.rs

```diff
@@ -0,0 +1,101 @@
+//! Tracer tracking deployment of EVM bytecodes during VM execution.
+
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
+
+use zk_evm_1_5_0::zkevm_opcode_defs::CALL_IMPLICIT_CALLDATA_FAT_PTR_REGISTER;
+use zksync_system_constants::{CONTRACT_DEPLOYER_ADDRESS, KNOWN_CODES_STORAGE_ADDRESS};
+use zksync_types::U256;
+use zksync_utils::{bytecode::hash_evm_bytecode, h256_to_u256};
+use zksync_vm2::{
+    interface::{
+        CallframeInterface, CallingMode, GlobalStateInterface, Opcode, OpcodeType, Tracer,
+    },
+    FatPointer,
+};
+
+/// Container for dynamic bytecodes added by [`EvmDeployTracer`].
+#[derive(Debug, Clone, Default)]
+pub(super) struct DynamicBytecodes(Rc<RefCell<HashMap<U256, Vec<u8>>>>);
+
+impl DynamicBytecodes {
+    pub(super) fn take(&self, hash: U256) -> Option<Vec<u8>> {
+        self.0.borrow_mut().remove(&hash)
+    }
+
+    fn insert(&self, hash: U256, bytecode: Vec<u8>) {
+        self.0.borrow_mut().insert(hash, bytecode);
+    }
+}
+
+/// Tracer that tracks EVM bytecode deployments.
+///
+/// Unlike EraVM bytecodes, EVM bytecodes are *dynamic*; they are not necessarily known before transaction execution.
+/// (EraVM bytecodes must be present in the storage or be mentioned in the `factory_deps` field of a transaction.)
+/// Hence, it's necessary to track which EVM bytecodes were deployed so that they are persisted after VM execution.
+#[derive(Debug)]
+pub(super) struct EvmDeployTracer {
+    tracked_signature: [u8; 4],
+    bytecodes: DynamicBytecodes,
+}
+
+impl EvmDeployTracer {
+    pub(super) fn new(bytecodes: DynamicBytecodes) -> Self {
+        let tracked_signature =
+            ethabi::short_signature("publishEVMBytecode", &[ethabi::ParamType::Bytes]);
+        Self {
+            tracked_signature,
+            bytecodes,
+        }
+    }
+
+    fn handle_far_call(&self, state: &mut impl GlobalStateInterface) {
+        let from = state.current_frame().caller();
+        let to = state.current_frame().code_address();
+        if from != CONTRACT_DEPLOYER_ADDRESS || to != KNOWN_CODES_STORAGE_ADDRESS {
+            return;
+        }
+
+        let (calldata_ptr, is_pointer) =
+            state.read_register(CALL_IMPLICIT_CALLDATA_FAT_PTR_REGISTER + 1);
+        assert!(
```

Comment:
I wouldn't assert things that are not critical for this function's correctness.

Reply:
I'd argue that in this case, both assertions are critical for correctness, in the sense that if they don't hold, the logic below doesn't work:

- If `r1` is not a pointer, this may mean a changed calling convention or a similar failure. The value in `r1` in this case cannot be interpreted.
- If `calldata_ptr.offset` is not zero, this means that the range of data to be read should change.

Would you prefer to allow for arbitrary `calldata_ptr.offset`? (I would be OK with this.) To return early if `r1` is not a pointer instead of panicking? (IMO, panicking is better because would allow to catch not updating calling convention logic etc. earlier.)

Reply:
My account validation PR adds a helper function `read_fat_pointer`, which does take offset into acount. I'd use that.

Switching to a different calling convention isn't possible, as it would break all existing bytecodes.